### ISSUE TITLE
Support `partition_key` on key within json field or key with dots

### ISSUE THREAD

**Issue by fmassot** [quickwit-oss/quickwit#3380]
Currently, it's not possible to define a partition key on
- a key in a JSON field
- a key that contains a dot (and probably containing special characters).

For example, if I create an index with a partition key on `span_attributes.tenant_id`, I get the following error:

```
Invalid config: Failed to interpret the partition key.: Parsing Error: ().
```

**Comment by jeromegn** [comment]
We'd be interested in using this.

I've had to remap some nested fields to be top level to make partitioning work.

**Comment by fulmicoton** [comment]
@jeromegn  Can you share the actual JSON? usually remapping is not necessary.

If your json looks like 
```json
{
    "my_obj": {
        "tenant_id" 33,
        "attribute1": "red",
        "attribute2": "red"
    }
}
```

We support declaring `my_obj.tenant_id` in your schema, and keeping the rest as dynamic.

**Comment by jeromegn** [comment]
I wasn't able to use a nested key for partitioning. I got the error described in the OP

**Comment by fulmicoton** [comment]
Understood!

**Comment by jeromegn** [comment]
Sorry, I wrote that as I was getting to bed with my brain fried.

The index looks like this:

```yaml
version: 0.6

index_id: app-logs

doc_mapping:
  field_mappings:
    - name: timestamp
      type: datetime
      precision: nanoseconds
      output_format: unix_timestamp_nanos
      fast: true
      
    - name: message
      type: text
      stored: true

    - name: fly
      type: object
      field_mappings:
      - name: org
        type: object
        field_mappings:
          - name: id
            type: u64
            fast: true
      - name: app
        type: object
        field_mappings:
          - name: id
            type: u64
            fast: true
          - name: name
            type: text
            tokenizer: raw
            fast: true
          - name: instance
            type: text
            tokenizer: raw
            fast: true
      - name: datacenter
        type: text
        tokenizer: raw
      - name: provider
        type: text
        tokenizer: raw
      - name: region
        type: text
        fast: true
        tokenizer: raw
    - name: host
      type: object
      field_mappings:
        - name: hostname
          type: text
          tokenizer: raw
    - name: event
      type: object
      field_mappings:
        - name: provider
          type: text
          tokenizer: raw
    - name: log
      type: object
      field_mappings:
        - name: level
          type: text
          tokenizer: raw

  tag_fields: []
  store_source: false
  timestamp_field: timestamp
  mode: dynamic
  partition_key: fly.org.id,fly.app.id

search_settings:
  default_search_fields: [message]

indexing_settings:
  commit_timeout_secs: 10
  split_num_docs_target: 10000000
  resources:
    heap_size: 2GB
```

This returns the same error as the original post.

**Comment by JerryKwan** [comment]
@fmassot @fulmicoton 
Seems like we should refactor these codes, and allow '.' in the identifier
https://github.com/quickwit-oss/quickwit/blob/main/quickwit/quickwit-doc-mapper/src/routing_expression/mod.rs#L370-L375

**Comment by fmassot** [comment]
Yes indeed. @trinity-1686a any thoughts on this?

**Comment by trinity-1686a** [comment]
that looks about right. `.` isn't currently used in the dsl so it should be relatively painless

**Comment by JerryKwan** [comment]
@fmassot @trinity-1686a 
I have tried to change identifier() to 
```rust
    fn identifier(input: &str) -> IResult<&str, &str> {
        println!("identifier() invoked, input = {}", input);
        input.split_at_position1_complete(
            |item| !(item.is_alphanum() || item == '_' || item == '.'),
            ErrorKind::AlphaNumeric,
        )
    }
    ```
And seems like it works. But I am not sure how to confirm if partition_key works as expected although create index and ingest documents all goes well
Any suggestions?

**Comment by trinity-1686a** [comment]
assuming your partition key is `abc.def`, you can ingest a bunch of documents `{"abc": {"def": "1"}}` `{"abc": {"def": "2"}}`...
Then go on quickwit UI, in Indexes > `your index name`, and see if you have as many split as unique values for `abc.def`. You can also check in the `splits` tabs (might not be shown if your viewport isn't wide enough), if splits have `tags` matching what you sent.
Finally you can do a search request which should use only a single split (`abc.def:1` for instance), search quickwit logs for a line starting with `root_search:root_search_aux:search_partial_hits_phase_with_scroll:search_partial_hits_phase:leaf_search`, and verify `splits_num=` in this line is 1.

**Comment by JerryKwan** [comment]
@trinity-1686a Thanks for your explanation.

**Comment by trinity-1686a** [comment]
fixed by #3989
